### PR TITLE
Refactor gateway app

### PR DESCRIPTION
**Pull request by harupy**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Refactors the gateway app to make it easier to test it.

## How is this patch tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [x] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/gateway/app.py [hunk 1]
```diff
@@ -109,34 +90,52 @@ def _route_type_to_endpoint(config: RouteConfig):
...
+            return {"route": matched}
 
-def create_app(gateway_conf_path: Union[str, Path]) -> FastAPI:
-    """
-    Create the FastAPI app by loading the dynamic route configuration file from the
-    specified local path and generating POST methods
-    """
-    route_config = _load_route_config(gateway_conf_path)
-    _add_routes(route_config.routes)
-    return app
```
Comment (harupy) [inline mlflow/gateway/app.py:133]: The original `create_app` was like a mutator. Every time we call it, it add new routes, which makes it difficult to test the app or separate tests.

File: mlflow/gateway/app.py [hunk 2]
```diff
@@ -2,63 +2,44 @@
...
-ACTIVE_ROUTES: List[Route] = []
-
-app = FastAPI(
-    title="MLflow Gateway API",
-    description="The core gateway API for reverse proxy interface using remote inference "
-    "endpoints within MLflow",
-    version=VERSION,
-)
 
+class GatewayApp(FastAPI):
```
Comment (harupy) [inline mlflow/gateway/app.py:25]: Encapsulate the dynamic route logic in this class.

File: mlflow/gateway/app.py [hunk 3]
```diff
@@ -2,63 +2,44 @@
...
-from .providers import get_provider
-from .schemas import chat, completions, embeddings
+from mlflow.gateway.providers import get_provider
 
 _logger = logging.getLogger(__name__)
 
 MLFLOW_GATEWAY_CONFIG = "MLFLOW_GATEWAY_CONFIG"
 
-# Configured and initialized Gateway Routes state index
-ACTIVE_ROUTES: List[Route] = []
```
Comment (harupy) [inline mlflow/gateway/app.py:25]: If we have a variable like this, we always have to be careful with mutation, so I removed it.

File: mlflow/gateway/app.py [hunk 4]
```diff
@@ -2,63 +2,44 @@
...
-        raise HTTPException(
-            status_code=404,
-            detail=f"The route '{route_name}' is not present or active on the server. Please "
-            "verify the route name.",
-        )
-    return {"route": filtered}
+class GatewayApp(FastAPI):
+    def __init__(self, config: GatewayConfig, *args: Any, **kwargs: Any):
+        super().__init__(*args, **kwargs)
+        self.dynamic_routes: Dict[str, Route] = {}
```
Comment (harupy) [inline mlflow/gateway/app.py:28]: Chose `dict` for faster look-up.

File: tests/gateway/test_gateway_app.py
Comment (harupy) [inline tests/gateway/test_gateway_app.py:1]: Split the test into smaller unit tests.

File: mlflow/gateway/app.py [hunk 5]
```diff
@@ -109,34 +90,55 @@ def _route_type_to_endpoint(config: RouteConfig):
...
-def _add_dynamic_route(route_config: RouteConfig):
-    app.add_api_route(
-        path=f"/gateway/routes/{route_config.name}",
-        endpoint=_route_type_to_endpoint(route_config),
-        methods=["POST"],
+def create_app_from_config(config: GatewayConfig) -> GatewayAPI:
+    """
+    Create the GatewayAPI app from the gateway configuration.
+    """
+    app = GatewayAPI(
```
Comment (BenWilson2) [inline mlflow/gateway/app.py:97]: This is excellent. +1 to the side effect elimination

**Review by BenWilson2 — APPROVED** [review]
LGTM! Great changes

**Pull request merged**